_posts/blog/2016-10-28-durability-performance.md

```diff
@@ -0,0 +1,136 @@
+`hsync` may take 50ms.  This difference will impact writes to Accumulo and can
+be mitigated in some situations with larger buffers in Accumulo.
+
+HDFS keeps checksum data internally by default.  Datanodes store checksum data
+in a separate file in the local filesystem.  This means when `hsync` is called
+on a WAL, two files must be synced on each datanode.  Syncing two files doubles
+the time. To make matters even worse, when the two files are synced the local
+filesystem metadata is also synced.  Depending on the local filesystem and its
+configuration, syncing the metadata may or may not take time.  In the worst
+case, we need to wait for four sync operations at the local filesystem level on
+each datanode. One thing I am not sure about, is if these sync operations occur
+in parallel on the replicas on different datanodes.  Lets hope they occur in
+parallel.  The following pointers show where sync occurs in the datanode code.
+
+ * [BlockReceiver.flushOrSync()][fos] calls [ReplicaOutputStreams.syncDataOut()][ros1] and [ReplicaOutputStreams.syncChecksumOut()][ros2] when `isSync` is true.
+ * The methods in ReplicaOutputStreams call [FileChannel.force(true)][fcf] which
+   synchronously flushes data and filesystem metadata.
+
+If files were preallocated (this would avoid syncing local filesystem metadata)
+and checksums were stored in-line, then 1 sync could be done instead of 4.  
+
+## Configuring WAL flush/sync in Accumulo 1.6
+
+Accumulo 1.6.0 only supported `hsync` and this caused [performance
+problems][160_RN_WAL].  In order to offer better performance, the option to
+configure `hflush` was [added in 1.6.1][161_RN_WAL].  The
+[tserver.wal.sync.method][16_UM_SM] configuration option was added to support
+this feature.  This was a tablet server wide option that applied to everything
+written to any table.   
+
+## Group Commit
+
+Each Accumulo Tablet Server has a single WAL.  When multiple clients send
+mutations to a tablet server at around the same time, the tablet sever may group
+all of this into a single WAL operation.  It will do this instead of writing and
+syncing or flushing each client's mutations to the WAL separately.  Doing this
+increase throughput and lowers average latency for clients.
+
+## Configuring WAL flush/sync in Accumulo 1.7+
+
+Accumulo 1.7.0 introduced [table.durability][17_UM_TD], a new per table property
+for configuring durability.  It also stopped using the `tserver.wal.sync.method`
+property.  The `table.durability` property has the following four legal values.
+
+ * **none** : Do not write to WAL            
+ * **log**  : Write to WAL, but do not sync  
+ * **flush** : Write to WAL and call `hflush` 
+ * **sync** : Write to WAL and call `hsync`  
+
+If multiple writes arrive at around the same time with different durability
+settings, then the group commit code will choose the most conservative
+durability.  This can cause one tables settings to slow down writes to another
+table.  
```

Comment:
This is very interesting. I hadn't considered this before. Sounds like this would be a very good metric to expose (since it could mutations written to one table much slower than expected).